### ISSUE TITLE
Parallelization of processing (I)

### ISSUE THREAD

**Issue by tesshucom** [tesshucom/jpsonic#1373]
Prerequisites: #1925.

In v111.x.x, processing is parallelized step by step. 

Eventually, speedup will be achieved, but if anything, there is a big refactoring aspect of replacing each processing block with appropriate processing.

<details>
<summary>Notes remaining in #833</summary>

There had been some points to be noted when reviewing existing threads.

1. Scalability : 
1. Subdivision of scanning process : 
1. Inappropriate exception handling
1. Other aggressive parallelization : the system statistics page, `@Async`

Of these, 1-3 will be improved with v111.x.x. 

As for 4, speeding up the web page is less important for finding and listening to music, so we can postpone it.

____

When developing a new system, you should start with architect design. It's a refactoring of an existing system in production, so to a large extent the reverse procedure is taken. The following approach will be taken.

#### Scalability : 

Safety is more important than speed when parallelizing existing systems. A mechanism that allows the user to adjust Scalability is essential.

Most of the current scan bottlenecks are not the speed of the database but IO control costs. Parallelization can alleviate some of this, but in most cases threading above the number of cores is counterproductive. Therefore, apart from the existing pooling, a new pooling that can adjust the size according to the number of cores will be newly installed, and a function will be added so that the user can control it.

Most modern CPUs support hyper-threading, so Jpsonic will provide logical threads / 2 as a recommended value(That is, maybe the number of cores). 

 - The user must optionally be able to assign a value lower than the recommended value.
 - User can assign a value higher than the recommended value. However, this is counterproductive in most cases and should be warned(This is only needed if the core does not support hyper-threading).

The JVM does not handle these things automatically. Systems that do not care about this mechanism will easily experience CPU highload issues. Parallelization should not be implemented as long as such users lead to uncontrollable problems.

#### Subdivision of scanning process / Inappropriate exception handling

It is difficult to fix all of these at once. This is because the range is too wide. Therefore, the theme will be set for each version, and improvements, tests, and effect measurements will be carried out for each block.

In v111.2.0, the brush-up around the parser, which has already been partially started, will be done. In addition, parsing is parallelized so that parts of the entire scan can be executed in parallel. The design of the newly created thread pool, queuing, and task is provisional, and if the verification and modification of other processing blocks are completed later, the task design may change cumulatively.

</details>

#### Fixes

Related bugs need to be fixed first.

- [x] #1376
- [x] #1392
- [x] #1405

v111.2.0
_____
v111.3.0

- [x] #1368
    -  The implementation is quite complicated, but the analysis time can be reduced to 1/10 or less in limited situations.
- [x] Add new fixed length thread pooling
- [ ] Add a feature to the page that allows users to change the number of concurrent scan tasks
- [ ] Fix parsing process so that it can be executed in parallel

**Comment by tesshucom** [comment]
- Add a feature to the page that allows users to change the number of concurrent scan tasks
- Fix parsing process so that it can be executed in parallel

These will be implemented in the future. In addition, materials necessary for design and implementation are becoming available.

 - #2310

However, the situation has changed considerably since then. Information about it will be codified in #1925.

Currently, Jpsonic has Docker installed, and many tunings are being performed using DS200+ as a verification machine. Now that most of the scans have been improved, v113 and v114 are the most important now.

 - v113 : The design of the index will be changed to include scans and the view will be improved. This is because it is the most expensive process that can be executed concurrently with Scan.
 - v114 : Podcasts will be redesigned and reimplemented. It is less practical. Also, it does not support canceling the process. Long-running Processes should be fully cancelable, like scans. When automatic service updates are provided, all functions should be designed with that in mind.

Therefore, it is better to give priority to v113 and v114, which have a higher degree of contribution to improving the stability of System.

And Also, since Java 21, virtual threads can be used. #2037. Haste makes waste.  Rather than being groundbreaking in itself, I would like information on Jakarta EE and Jetty. If we digest it in a good order, there is a possibility that the gears will turn well.
